<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="smarterapi">
<title>Using smarterapi • smarterapi</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Using smarterapi">
<meta property="og:description" content="smarterapi">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">smarterapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/smarterapi.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/geojson.html">Working with GeoJSON data</a>
    <a class="dropdown-item" href="../articles/variants.html">Working with SNPs</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cnr-ibba/r-smarter-api/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Using smarterapi</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cnr-ibba/r-smarter-api/blob/HEAD/vignettes/smarterapi.Rmd" class="external-link"><code>vignettes/smarterapi.Rmd</code></a></small>
      <div class="d-none name"><code>smarterapi.Rmd</code></div>
    </div>

    
    
<p><strong>smarterapi</strong> is an R package which enable to browse
the <a href="https://webserver.ibba.cnr.it/smarter-api/docs/" class="external-link">SMARTER
backend API</a> and collect smarter data. In this R package, functions
to deal with the different API endpoints are implemented.</p>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Simply import <code>smarterapi</code> like any other R package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cnr-ibba.github.io/r-smarter-api/">smarterapi</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="collect-info-about-versions">Collect info about versions<a class="anchor" aria-label="anchor" href="#collect-info-about-versions"></a>
</h2>
<p>The smarter database is continuously updated, for example when
receiving new samples, genotypes, metadata or when something is fixed or
updated, for example when a new assembly is added. In order to ensure
reproducibility and monitoring changes, genotypes and metadata are
provided using <em>versions</em>: this means that you have to ensure
that the genotypes dataset and metadata are referring to the same
version. You can check database version using the
<code><a href="../reference/get_smarter_info.html">get_smarter_info()</a></code> method:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_info.html">get_smarter_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">info</span><span class="op">$</span><span class="va">version</span></span>
<span><span class="co">#&gt; [1] "0.4.10"</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">info</span><span class="op">$</span><span class="va">last_updated</span></span>
<span><span class="co">#&gt; [1] "2024-05-29 02:41:35 UTC"</span></span></code></pre></div>
<p>The same version system is specified in the genotype file dataset you
can download from the smarter FTP site:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">$</span> tree <span class="at">-L</span> 2 SHEEP/</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">SHEEP/</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="ex">└──</span> OAR3</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>    <span class="ex">├──</span> archive</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>    <span class="ex">├──</span> SMARTER-OA-OAR3-top-0.4.10.md5</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>    <span class="ex">└──</span> SMARTER-OA-OAR3-top-0.4.10.zip</span></code></pre></div>
<p>You have to ensure that you are working with the same version, for
both metadata and genotypes: if not, you may not find samples in the
genotype files or you are referring to an out-dated assembly version. To
see the changelist in each version, please refer to the SMARTER-database
<a href="https://github.com/cnr-ibba/SMARTER-database/blob/master/HISTORY.rst" class="external-link">HISTORY.rst</a>
file.</p>
</div>
<div class="section level2">
<h2 id="querying-smarter-api">Querying SMARTER API<a class="anchor" aria-label="anchor" href="#querying-smarter-api"></a>
</h2>
<p>The <code>smarterapi</code> package is structured to handle
<em>requests/responses</em> to the SMARTER API backend using R packages
like <code>httr</code> and <code>jsonlite</code>. In general, each
functions accepts a <code>query=list()</code> parameter, in which pass
additional parameter to the API endpoint in order to filter results
matching query. Some functions like <code><a href="../reference/get_smarter_samples.html">get_smarter_samples()</a></code>
or <code><a href="../reference/get_smarter_variants.html">get_smarter_variants()</a></code> requires additional parameters
like the <code>species</code> or the <code>assembly</code> version. Each
function in general will return results in a <code>data.frame</code>:
you could filter out data using <code>dplyr</code> or with standard R
methods or you could filter data <em>directly</em> by submitting a
proper query to the API. Please refer to the proper documentation to
understand which parameters a function expect. See also the <a href="https://webserver.ibba.cnr.it/smarter-api/docs/" class="external-link">SMARTER-backend
API</a> web interface to have an idea on which parameter are allowed for
each endpoint.</p>
</div>
<div class="section level2">
<h2 id="collect-samples">Collect samples<a class="anchor" aria-label="anchor" href="#collect-samples"></a>
</h2>
<p>Here are some examples on collecting samples. The main function
required to collect sample is <code><a href="../reference/get_smarter_samples.html">get_smarter_samples()</a></code>, which
is an helper function which allow to query the <a href="https://webserver.ibba.cnr.it/smarter-api/docs/#/Samples/get_samples_goat" class="external-link">/samples/goat</a>
and <a href="https://webserver.ibba.cnr.it/smarter-api/docs/#/Samples/get_samples_sheep" class="external-link">/samples/sheep</a>
endpoints of the SMARTER-backend API. We will use some other functions
to have better idea on which values to use to filter data. Remember that
all the parameters you see in each different example can be merged with
other to restrict your query to have only the samples you are looking
for.</p>
<div class="section level3">
<h3 id="select-by-datasets">Select by datasets<a class="anchor" aria-label="anchor" href="#select-by-datasets"></a>
</h3>
<p>Getting samples by dataset is quite easy, you have to provide the
proper <code>dataset_id</code> to the <code><a href="../reference/get_smarter_samples.html">get_smarter_samples()</a></code>
function. However, you have to determine the proper
<code>dataset_id</code> using the <code>get_smarter_dataset()</code>
function, which model the <a href="https://webserver.ibba.cnr.it/smarter-api/docs/#/Datasets/get_datasets" class="external-link">/datasets</a>
endpoint. For example, to extract only <em>background</em> genotypes
datasets (data generated before SMARTER project) you have to pass the
proper <code>type</code> to the <code>query</code> argument. Since the
<code>query</code> argument is a list, you can pass multiple parameters
at once. For parameters which supports arrays, you could supply the same
parameter name multiple times: each one will be passed through the API
endpoint</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select all genotypes datasets made before SMARTER project</span></span>
<span><span class="va">background_genotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_datasets.html">get_smarter_datasets</a></span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"background"</span>, type <span class="op">=</span> <span class="st">"genotypes"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># same as before, but limiting to goat species</span></span>
<span><span class="va">background_goat_genotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_datasets.html">get_smarter_datasets</a></span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="st">"background"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"genotypes"</span>,</span>
<span>    species <span class="op">=</span> <span class="st">"Goat"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Take some time to explore the dataframe columns. There are two
importants fields, the first is the <code>_id.$oid</code> column, which
is the <code>dataset_id</code> we want to provide to collect samples
belonging to this dataset. The second is the <code>file</code> column,
which is the archive name which was uploaded into the smarter database.
For example, here is what the <code>background_goat_genotypes</code>
table looks like:</p>
<table class="table">
<colgroup>
<col width="35%">
<col width="17%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th align="center">_id.$oid</th>
<th align="center">breed</th>
<th align="center">file</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">604f75a61a08c53cebd09b5b</td>
<td align="center">144 breeds</td>
<td align="center">ADAPTmap_genotypeTOP_20161201.zip</td>
</tr>
<tr class="even">
<td align="center">643ec41371506d0c63bc5d63</td>
<td align="center">10 breeds</td>
<td align="center">burren_et_al_2016.zip</td>
</tr>
<tr class="odd">
<td align="center">643ec41571506d0c63bc5d64</td>
<td align="center">23 breeds</td>
<td align="center">cortellari_et_al_2021.zip</td>
</tr>
</tbody>
</table>
<p>So collect the <em>adaptmap</em> samples, we can provide the proper
<code>dataset_id</code> to the <code><a href="../reference/get_smarter_samples.html">get_smarter_samples()</a></code>
method. We can add additional parameters, like country:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select the adaptmap id, which is in the first row of the dataframe</span></span>
<span><span class="va">adatpmap_id</span> <span class="op">&lt;-</span> <span class="va">background_goat_genotypes</span><span class="op">[</span><span class="st">"_id.$oid"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">1</span><span class="op">]</span></span>
<span><span class="va">adaptmap_goats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_samples.html">get_smarter_samples</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Goat"</span>, query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="va">adatpmap_id</span>, country <span class="op">=</span> <span class="st">"Italy"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>The previous case is quite easy, we want only one dataset in
<code>background_goat_genotypes</code> dataframe, so we can simply paste
this value in the <code>get_smarter_samples</code> query. But how we can
handle multiple datasets? we can transform the proper column in a list
and then renaming it:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get more datasets</span></span>
<span><span class="va">foreground_goat_genotypes</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_datasets.html">get_smarter_datasets</a></span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"genotypes"</span>, type <span class="op">=</span> <span class="st">"foreground"</span>, species <span class="op">=</span> <span class="st">"Goat"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># construct the query arguments</span></span>
<span><span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/as.list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">foreground_goat_genotypes</span><span class="op">$</span><span class="st">"_id.$oid"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">datasets</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"dataset"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">datasets</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">breeds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>breed_code <span class="op">=</span> <span class="st">"LNR"</span>, breed_code <span class="op">=</span> <span class="st">"SKO"</span>, breed_code <span class="op">=</span> <span class="st">"FSS"</span><span class="op">)</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">datasets</span>, <span class="va">breeds</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select samples: subset by breed code and datasets</span></span>
<span><span class="va">foreground_goat_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_samples.html">get_smarter_samples</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Goat"</span>, query <span class="op">=</span> <span class="va">query</span><span class="op">)</span></span></code></pre></div>
<p>We can also use the <code><a href="../reference/get_smarter_datasets.html">get_smarter_datasets()</a></code> function to
do a reverse selection of our samples, for example to get all the
samples which are not in one or more datasets: suppose, for example, to
collect all samples which are not in the <em>isheep</em> datasets:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># collect all foreground datasets having "isheep" in their name</span></span>
<span><span class="va">isheep_datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_datasets.html">get_smarter_datasets</a></span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    type <span class="op">=</span> <span class="st">"foreground"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"genotypes"</span>,</span>
<span>    search <span class="op">=</span> <span class="st">"isheep"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># collect ids for isheep datasets</span></span>
<span><span class="va">isheep_ids</span> <span class="op">&lt;-</span> <span class="va">isheep_datasets</span><span class="op">$</span><span class="st">"_id.$oid"</span></span>
<span></span>
<span><span class="co"># collect all foreground samples</span></span>
<span><span class="va">foreground_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_samples.html">get_smarter_samples</a></span><span class="op">(</span></span>
<span>  <span class="st">"Sheep"</span>, query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"foreground"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get rid of isheep_samples</span></span>
<span><span class="va">filtered_samples</span> <span class="op">&lt;-</span> <span class="va">foreground_samples</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">`dataset_id.$oid`</span> <span class="op"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">%in%</a></span> <span class="va">isheep_ids</span><span class="op">)</span></span></code></pre></div>
<p>The last selection example relies on dataset file contents: if you
remember the name of the file submitted in the dataset, you can search
by datasets <em>content</em>:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_datasets.html">get_smarter_datasets</a></span><span class="op">(</span>query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>search <span class="op">=</span> <span class="st">"adaptmap"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="35%">
<col width="17%">
<col width="47%">
</colgroup>
<thead><tr class="header">
<th align="center">_id.$oid</th>
<th align="center">breed</th>
<th align="center">file</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">604f75a61a08c53cebd09b5b</td>
<td align="center">144 breeds</td>
<td align="center">ADAPTmap_genotypeTOP_20161201.zip</td>
</tr>
<tr class="even">
<td align="center">604f82821a08c53cebd0a0bf</td>
<td align="center">144 breeds</td>
<td align="center">ADAPTmap_phenotype_20161201.zip</td>
</tr>
</tbody>
</table>
<p>This time two results are returned, since one is a
<em>phenotypes</em> dataset, while the other is a <em>genotypes</em>. To
select only <em>genotypes</em>, simply add <code>type=genotypes</code>
to the <code>query</code> parameter.</p>
</div>
<div class="section level3">
<h3 id="select-by-breed">Select by breed<a class="anchor" aria-label="anchor" href="#select-by-breed"></a>
</h3>
<p>You can select samples relying on <em>breeds names</em> or <em>breed
codes</em>. Breed names are written in the languages they come from, so
in order to retrieve <em>Île de France</em> or <em>Fjällnäs</em> breed
samples, you have to specify the full breed name or use the search
parameter with the <code><a href="../reference/get_smarter_breeds.html">get_smarter_breeds()</a></code> which model the <a href="https://webserver.ibba.cnr.it/smarter-api/docs/#/Breeds/get_breeds" class="external-link">/breeds</a>
endpoint:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">breeds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_breeds.html">get_smarter_breeds</a></span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    species <span class="op">=</span> <span class="st">"Sheep"</span>,</span>
<span>    search <span class="op">=</span> <span class="st">"de france"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<table style="width:32%;" class="table">
<colgroup>
<col width="22%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="center">name</th>
<th align="center">code</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center">Île de France</td>
<td align="center">IDF</td>
</tr></tbody>
</table>
<p>Search for breeds can return multiple values, for example:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">breeds</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_breeds.html">get_smarter_breeds</a></span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    species <span class="op">=</span> <span class="st">"Sheep"</span>,</span>
<span>    search <span class="op">=</span> <span class="st">"merino"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<table style="width:47%;" class="table">
<colgroup>
<col width="37%">
<col width="9%">
</colgroup>
<thead><tr class="header">
<th align="center">name</th>
<th align="center">code</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">Merino</td>
<td align="center">MER</td>
</tr>
<tr class="even">
<td align="center">AustralianIndustryMerino</td>
<td align="center">AIM</td>
</tr>
<tr class="odd">
<td align="center">AustralianMerino</td>
<td align="center">AME</td>
</tr>
<tr class="even">
<td align="center">AustralianPollMerino</td>
<td align="center">APM</td>
</tr>
<tr class="odd">
<td align="center">ChineseMerino</td>
<td align="center">CME</td>
</tr>
<tr class="even">
<td align="center">MacarthurMerino</td>
<td align="center">MCM</td>
</tr>
<tr class="odd">
<td align="center">Merinolandschaf</td>
<td align="center">MLA</td>
</tr>
<tr class="even">
<td align="center">Merino Horned</td>
<td align="center">MEH</td>
</tr>
<tr class="odd">
<td align="center">Merino Polled</td>
<td align="center">MEP</td>
</tr>
<tr class="even">
<td align="center">Merino Estremadura</td>
<td align="center">MEE</td>
</tr>
</tbody>
</table>
<p>Name and codes can be used as they are to select samples by passing
multiple arguments to the query:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_samples.html">get_smarter_samples</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Sheep"</span>, query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  breed_code <span class="op">=</span> <span class="st">"MER"</span>, breed_code <span class="op">=</span> <span class="st">"AME"</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>or to get all the samples with <em>merino</em> in breed name:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># construct the query arguments</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/as.list.html" class="external-link">as.list</a></span><span class="op">(</span><span class="va">breeds</span><span class="op">$</span><span class="va">code</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"breed_code"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">query</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># execute query</span></span>
<span><span class="va">merino_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_samples.html">get_smarter_samples</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Sheep"</span>, query <span class="op">=</span> <span class="va">query</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="select-by-country">Select by country<a class="anchor" aria-label="anchor" href="#select-by-country"></a>
</h3>
<p>You can retrive samples by countries. First get a list of the
available countries relying on country name, then extract samples using
the correct country name:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">italy</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_countries.html">get_smarter_countries</a></span><span class="op">(</span>query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>search <span class="op">=</span> <span class="st">"italy"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">italian_background_sheeps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_samples.html">get_smarter_samples</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Sheep"</span>,</span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    country <span class="op">=</span> <span class="va">italy</span><span class="op">$</span><span class="va">name</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="select-by-chip">Select by chip<a class="anchor" aria-label="anchor" href="#select-by-chip"></a>
</h3>
<p>You can select samples relying on the chip they are sequenced. If you
search for multiple chip types, you will collect all samples which
belongs to any of the specified chip. First, collect a list of the
available chips for a certain species:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sheep_chips</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_supportedchips.html">get_smarter_supportedchips</a></span><span class="op">(</span>query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Sheep"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table style="width:88%;" class="table">
<colgroup>
<col width="20%">
<col width="16%">
<col width="36%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="center">manufacturer</th>
<th align="center">n_of_snps</th>
<th align="center">name</th>
<th align="center">species</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">illumina</td>
<td align="center">54241</td>
<td align="center">IlluminaOvineSNP50</td>
<td align="center">Sheep</td>
</tr>
<tr class="even">
<td align="center">illumina</td>
<td align="center">605998</td>
<td align="center">IlluminaOvineHDSNP</td>
<td align="center">Sheep</td>
</tr>
<tr class="odd">
<td align="center">affymetrix</td>
<td align="center">56793</td>
<td align="center">AffymetrixAxiomOviCan</td>
<td align="center">Sheep</td>
</tr>
<tr class="even">
<td align="center">affymetrix</td>
<td align="center">49702</td>
<td align="center">AffymetrixAxiomBGovisNP</td>
<td align="center">Sheep</td>
</tr>
<tr class="odd">
<td align="center">affymetrix</td>
<td align="center">60379</td>
<td align="center">AffymetrixAxiomBGovis2</td>
<td align="center">Sheep</td>
</tr>
<tr class="even">
<td align="center">NA</td>
<td align="center">0</td>
<td align="center">WholeGenomeSequencing</td>
<td align="center">Sheep</td>
</tr>
</tbody>
</table>
<p>Then collect samples relying on chip name, for example:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_samples.html">get_smarter_samples</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Sheep"</span>,</span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    chip_name <span class="op">=</span> <span class="st">"IlluminaOvineHDSNP"</span>,</span>
<span>    chip_name <span class="op">=</span> <span class="st">"AffymetrixAxiomOviCan"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="select-by-metadata">Select by metadata<a class="anchor" aria-label="anchor" href="#select-by-metadata"></a>
</h3>
<p>Since metadata aren’t formatted in the same way in each samples, is
difficult to define a single query you can apply to each samples. For
the moment, the only queries you can apply on metadata are restricted to
their presence or absence. For example, we can collect all samples which
have GPS coordinates and phenotypes (any):</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">smarter_goats</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_samples.html">get_smarter_samples</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Goat"</span>,</span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    locations__exists <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    phenotype__exists <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>After that, you have to filter out the <code>smarter_goats</code>
dataframe in order to collect only the samples you want.</p>
</div>
</div>
<div class="section level2">
<h2 id="refer-samples-to-their-original-publication">Refer samples to their original publication<a class="anchor" aria-label="anchor" href="#refer-samples-to-their-original-publication"></a>
</h2>
<p>Suppose you want to refer the samples to the original publication the
come from: in the <em>dataset</em> dataframe we have a <code>doi</code>
column in which the <em>publication DOI</em> is stored. You can use this
information to collect publication information, but you need to merge
<em>samples</em> and <em>datasets</em> dataframes in order to refer
properly the samples to the publication they come from. Here is an
example of how to do:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># collect all the genotypes datasets</span></span>
<span><span class="va">datasets</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_datasets.html">get_smarter_datasets</a></span><span class="op">(</span></span>
<span>  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    species <span class="op">=</span> <span class="st">"Sheep"</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"genotypes"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># collect all the samples</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_samples.html">get_smarter_samples</a></span><span class="op">(</span></span>
<span>  species <span class="op">=</span> <span class="st">"Sheep"</span>,</span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># merge datasets and samples using dplyr</span></span>
<span><span class="va">samples_with_doi</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span></span>
<span>  <span class="va">datasets</span>,</span>
<span>  <span class="va">samples</span>,</span>
<span>  by <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/join_by.html" class="external-link">join_by</a></span><span class="op">(</span><span class="va">`_id.$oid`</span> <span class="op">==</span> <span class="va">`dataset_id.$oid`</span><span class="op">)</span></span>
<span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">smarter_id</span>, <span class="va">breed_code</span>, <span class="va">breed.y</span>, <span class="va">doi</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">doi</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># count how many samples come from each publication</span></span>
<span><span class="va">samples_with_doi</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">doi</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="collect-genotypes">Collect genotypes<a class="anchor" aria-label="anchor" href="#collect-genotypes"></a>
</h2>
<p>The <em>genotypes</em> can’t be retrieved using the
<em>smarter-api</em> because they are available from the <a href="ftp://webserver.ibba.cnr.it/smarter/">public FTP</a>. You need to
download the files using an FTP client like <a href="https://filezilla-project.org/" class="external-link">FileZilla</a> or <a href="https://lftp.yar.ru/" class="external-link">lftp</a> and then extract the genotypes
using <a href="https://www.cog-genomics.org/plink/" class="external-link">plink</a>. In
alternative, is it possible to download the genotypes using the
<code>get_smarter_genotypes</code> method of this package. This method
will download the genotypes from the FTP for the current releases of the
selected <em>species</em> and <em>assembly</em> and will return the
destination path of the downloaded archive:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">downloaded_archive</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_genotypes.html">get_smarter_genotypes</a></span><span class="op">(</span></span>
<span>  <span class="st">"Sheep"</span>,</span>
<span>  <span class="st">"OAR3"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This method will download the genotypes in the current working
directory, or in the directory specified in the <code>dest_path</code>
argument. The genotypes will be stored in a compressed <code>.zip</code>
file, which need to be de-compressed in order to be used with <a href="https://www.cog-genomics.org/plink/" class="external-link">plink</a>.</p>
</div>
<div class="section level2">
<h2 id="subset-genotypes-relying-samples">Subset genotypes relying samples<a class="anchor" aria-label="anchor" href="#subset-genotypes-relying-samples"></a>
</h2>
<p>After you have identified the samples of your interest, you can
extract their genotypes from the proper file using <a href="https://www.cog-genomics.org/plink/" class="external-link">plink</a>. First, you have to
write a <em>TSV</em> file with <em>breed_code</em> and
<em>smarter_id</em> as columns. For example, using the samples selected
above and <em>dplyr</em>:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">selected_sheeps_ids</span> <span class="op">&lt;-</span> <span class="va">italian_background_sheeps</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>    <span class="st">"breed_code"</span>, <span class="st">"smarter_id"</span></span>
<span>  <span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span></span>
<span>  <span class="va">selected_sheeps_ids</span>,</span>
<span>  file <span class="op">=</span> <span class="st">"selected_sheeps.txt"</span>,</span>
<span>  quote <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  sep <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>  row.names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  col.names <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Next, you need to collect the proper <em>plink options</em> in order
to not loose information from the plink file. The parameters used to
generate the genotype files are tracked in the info endpoint. In this
example, get parameters from <em>Sheep</em> genotypes:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_info.html">get_smarter_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">plink_opts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">info</span><span class="op">$</span><span class="va">plink_specie_opt</span><span class="op">$</span><span class="va">Sheep</span>, collapse <span class="op">=</span> <span class="st">" "</span><span class="op">)</span></span>
<span><span class="va">plink_opts</span></span>
<span><span class="co">#&gt; [1] "--chr-set 26 no-xy no-mt --allow-no-sex"</span></span></code></pre></div>
<p>And finally you can call plink and providing your sample list:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="ex">plink</span> <span class="at">--chr-set</span> 26 no-xy no-mt <span class="at">--allow-no-sex</span> <span class="dt">\</span></span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a>  <span class="at">--bfile</span> SMARTER-OA-OAR3-top-0.4.10 <span class="dt">\</span></span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a>  <span class="at">--keep</span> selected_sheeps.txt <span class="dt">\</span></span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a>  <span class="at">--out</span> selected_sheeps-OAR3-top-0.4.10 <span class="dt">\</span></span>
<span id="cb22-5"><a href="#cb22-5" tabindex="-1"></a>  <span class="at">--make-bed</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Paolo Cozzi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
