<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="smarterapi">
<title>Working with SNPs • smarterapi</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.1.0/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.1.0/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Working with SNPs">
<meta property="og:description" content="smarterapi">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">smarterapi</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.2</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/smarterapi.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/geojson.html">Working with GeoJSON data</a>
    <a class="dropdown-item" href="../articles/variants.html">Working with SNPs</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cnr-ibba/r-smarter-api/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Working with SNPs</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/cnr-ibba/r-smarter-api/blob/HEAD/vignettes/variants.Rmd" class="external-link"><code>vignettes/variants.Rmd</code></a></small>
      <div class="d-none name"><code>variants.Rmd</code></div>
    </div>

    
    
<p>This vignette describe how to use the <a href="https://webserver.ibba.cnr.it/smarter-api/docs/#/Variants" class="external-link">variant endpoints</a> which store information about SNPs used in the <em>smarter genotype datasets</em>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://cnr-ibba.github.io/r-smarter-api/">smarterapi</a></span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="assembly-versions">Assembly versions<a class="anchor" aria-label="anchor" href="#assembly-versions"></a>
</h2>
<p>One of the aim of this project is to manage genotypes in different assembly version. This means collect data from different assemblies (due to when data is generated), from different sources (<em>Affymetrix</em>, <em>Illumina</em>, <em>WGS</em>) and different file formats. Genotypes are normalized in order to be consistent accross data sources and stored in one genotype file for each specie.</p>
<p>Currently four assembly versions are managed, two for the sheep dataset and two for the goat dataset. Information about assemblies data sources can be retrieved from the backend <a href="https://webserver.ibba.cnr.it/smarter-api/docs/#/Info" class="external-link">info endpoint</a> through the <code><a href="../reference/get_smarter_info.html">get_smarter_info()</a></code> function:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_info.html">get_smarter_info</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">assemblies</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/transpose.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/terra/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">info</span><span class="op">$</span><span class="va">working_assemblies</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/terra/man/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">assemblies</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/terra/man/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"name"</span>, <span class="st">"source"</span><span class="op">)</span></code></pre></div>
<table style="width:51%;" class="table">
<colgroup>
<col width="15%">
<col width="15%">
<col width="20%">
</colgroup>
<thead><tr class="header">
<th align="center"> </th>
<th align="center">name</th>
<th align="center">source</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><strong>ARS1</strong></td>
<td align="center">ARS1</td>
<td align="center">manifest</td>
</tr>
<tr class="even">
<td align="center"><strong>CHI1</strong></td>
<td align="center">CHI1.0</td>
<td align="center">SNPchiMp v.3</td>
</tr>
<tr class="odd">
<td align="center"><strong>OAR3</strong></td>
<td align="center">Oar_v3.1</td>
<td align="center">SNPchiMp v.3</td>
</tr>
<tr class="even">
<td align="center"><strong>OAR4</strong></td>
<td align="center">Oar_v4.0</td>
<td align="center">SNPchiMp v.3</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="collect-data-from-an-assembly">Collect data from an assembly<a class="anchor" aria-label="anchor" href="#collect-data-from-an-assembly"></a>
</h2>
<p><code><a href="../reference/get_smarter_variants.html">get_smarter_variants()</a></code> have two mandatory parameters, <code>species</code> and <code>assembly</code>, then it could accept additional parameters (see one of the <a href="https://webserver.ibba.cnr.it/smarter-api/docs/#/Variants" class="external-link">variant endpoints</a> to have more information). For example you can search variants for <em>snp name</em> or <em>rs id</em> (if the latter exists):</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_variants.html">get_smarter_variants</a></span><span class="op">(</span>
  species <span class="op">=</span> <span class="st">"Goat"</span>, 
  assembly <span class="op">=</span> <span class="st">"ARS1"</span>,
  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
    name <span class="op">=</span> <span class="st">"snp12965-scaffold1499-3295573"</span>
  <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<table style="width:97%;" class="table">
<caption>Table continues below</caption>
<colgroup>
<col width="27%">
<col width="11%">
<col width="18%">
<col width="15%">
<col width="25%">
</colgroup>
<thead><tr class="header">
<th align="center">chip_name</th>
<th align="center">chrom</th>
<th align="center">date</th>
<th align="center">illumina</th>
<th align="center">illumina_strand</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center">IlluminaGoatSNP50</td>
<td align="center">8</td>
<td align="center">2021-01-08</td>
<td align="center">T/C</td>
<td align="center">BOT</td>
</tr></tbody>
</table>
<table style="width:85%;" class="table">
<caption>Table continues below</caption>
<colgroup>
<col width="20%">
<col width="22%">
<col width="15%">
<col width="12%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="center">illumina_top</th>
<th align="center">imported_from</th>
<th align="center">position</th>
<th align="center">strand</th>
<th align="center">version</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center">A/G</td>
<td align="center">manifest</td>
<td align="center">66621364</td>
<td align="center">BOT</td>
<td align="center">ARS1</td>
</tr></tbody>
</table>
<table style="width:64%;" class="table">
<colgroup>
<col width="44%">
<col width="19%">
</colgroup>
<thead><tr class="header">
<th align="center">name</th>
<th align="center">rs_id</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center">snp12965-scaffold1499-3295573</td>
<td align="center">rs119103301</td>
</tr></tbody>
</table>
<p>Please, refer to the <code><a href="../reference/get_smarter_info.html">get_smarter_info()</a></code> <code>working_assemblies</code> to have an idea of the assemblies supported by the <a href="https://github.com/cnr-ibba/SMARTER-database" class="external-link">SMARTER-database</a>. Data which come from <a href="https://webserver.ibba.cnr.it/SNPchimp/" class="external-link">SNPchiMp v.3</a> like the <em>Sheep</em> <code>OAR3</code> assembly, support the <em>illumina forward</em> attribute. For example the following SNP:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_variants.html">get_smarter_variants</a></span><span class="op">(</span>
  species <span class="op">=</span> <span class="st">"Sheep"</span>, 
  assembly <span class="op">=</span> <span class="st">"OAR3"</span>,
  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
    rs_id <span class="op">=</span> <span class="st">"rs10721092"</span>
  <span class="op">)</span>
<span class="op">)</span></code></pre></div>
<table style="width:96%;" class="table">
<caption>Table continues below</caption>
<colgroup>
<col width="29%">
<col width="13%">
<col width="11%">
<col width="15%">
<col width="26%">
</colgroup>
<thead><tr class="header">
<th align="center">chip_name</th>
<th align="center">alleles</th>
<th align="center">chrom</th>
<th align="center">illumina</th>
<th align="center">illumina_forward</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center">IlluminaOvineHDSNP</td>
<td align="center">C/T</td>
<td align="center">18</td>
<td align="center">T/C</td>
<td align="center">T/C</td>
</tr></tbody>
</table>
<table class="table">
<caption>Table continues below</caption>
<colgroup>
<col width="24%">
<col width="20%">
<col width="21%">
<col width="14%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th align="center">illumina_strand</th>
<th align="center">illumina_top</th>
<th align="center">imported_from</th>
<th align="center">position</th>
<th align="center">ss_id</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center">bottom</td>
<td align="center">A/G</td>
<td align="center">SNPchiMp v.3</td>
<td align="center">64249211</td>
<td align="center">ss896246690</td>
</tr></tbody>
</table>
<table style="width:78%;" class="table">
<colgroup>
<col width="13%">
<col width="15%">
<col width="30%">
<col width="18%">
</colgroup>
<thead><tr class="header">
<th align="center">strand</th>
<th align="center">version</th>
<th align="center">name</th>
<th align="center">rs_id</th>
</tr></thead>
<tbody><tr class="odd">
<td align="center">forward</td>
<td align="center">Oar_v3.1</td>
<td align="center">oar3_OAR18_64249211</td>
<td align="center">rs10721092</td>
</tr></tbody>
</table>
<p>Is <code>T/C</code> on the <em>forward</em> strand of <em>OAR3</em>: this means that the reversed probe is aligned to the genome (as you could infer from the <code>bottom</code> <em>illumina strand</em> attribute of this SNP.). Variants in the SMARTER-database are converted using the <strong>illumina top</strong> coding convenction, so you will find this SNP as <code>A/G</code> in the SMARTER-database while on the reference sequence it’s <code>T/C</code>.</p>
</div>
<div class="section level2">
<h2 id="fetch-variants-by-region">Fetch Variants by region<a class="anchor" aria-label="anchor" href="#fetch-variants-by-region"></a>
</h2>
<p>Variants endpoint support query by regions, using <code>&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;</code> as format, for example:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_variants.html">get_smarter_variants</a></span><span class="op">(</span>
  species <span class="op">=</span> <span class="st">"Goat"</span>,
  assembly <span class="op">=</span> <span class="st">"ARS1"</span>,
  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>region <span class="op">=</span> <span class="st">"1:1-100000"</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<table style="width:97%;" class="table">
<caption>Table continues below</caption>
<colgroup>
<col width="27%">
<col width="11%">
<col width="18%">
<col width="15%">
<col width="25%">
</colgroup>
<thead><tr class="header">
<th align="center">chip_name</th>
<th align="center">chrom</th>
<th align="center">date</th>
<th align="center">illumina</th>
<th align="center">illumina_strand</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">IlluminaGoatSNP50</td>
<td align="center">1</td>
<td align="center">2021-01-08</td>
<td align="center">T/C</td>
<td align="center">BOT</td>
</tr>
<tr class="even">
<td align="center">IlluminaGoatSNP50</td>
<td align="center">1</td>
<td align="center">2021-01-08</td>
<td align="center">T/C</td>
<td align="center">BOT</td>
</tr>
<tr class="odd">
<td align="center">IlluminaGoatSNP50</td>
<td align="center">1</td>
<td align="center">2021-01-08</td>
<td align="center">T/C</td>
<td align="center">BOT</td>
</tr>
</tbody>
</table>
<table style="width:85%;" class="table">
<caption>Table continues below</caption>
<colgroup>
<col width="20%">
<col width="22%">
<col width="15%">
<col width="12%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="center">illumina_top</th>
<th align="center">imported_from</th>
<th align="center">position</th>
<th align="center">strand</th>
<th align="center">version</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">A/G</td>
<td align="center">manifest</td>
<td align="center">18960</td>
<td align="center">TOP</td>
<td align="center">ARS1</td>
</tr>
<tr class="even">
<td align="center">A/G</td>
<td align="center">manifest</td>
<td align="center">47271</td>
<td align="center">BOT</td>
<td align="center">ARS1</td>
</tr>
<tr class="odd">
<td align="center">A/G</td>
<td align="center">manifest</td>
<td align="center">63095</td>
<td align="center">BOT</td>
<td align="center">ARS1</td>
</tr>
</tbody>
</table>
<table style="width:62%;" class="table">
<colgroup>
<col width="43%">
<col width="19%">
</colgroup>
<thead><tr class="header">
<th align="center">name</th>
<th align="center">rs_id</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">1_18960_AF-PAKI</td>
<td align="center">NA</td>
</tr>
<tr class="even">
<td align="center">snp14099-scaffold1560-920888</td>
<td align="center">rs268246860</td>
</tr>
<tr class="odd">
<td align="center">GoatD01.029677</td>
<td align="center">NA</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="fetch-variants-by-chip-name">Fetch Variants by chip name<a class="anchor" aria-label="anchor" href="#fetch-variants-by-chip-name"></a>
</h2>
<p>You can download <em>all variants</em> for a certain chip: please consider that it will require a lot of time and memory, since we store more than 600K SNPs in the smarter database. First, collect the available chips from the SMARTER-database, for example for the Sheep species:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sheep_chips</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_supportedchips.html">get_smarter_supportedchips</a></span><span class="op">(</span>query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"Sheep"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<table style="width:85%;" class="table">
<colgroup>
<col width="20%">
<col width="16%">
<col width="33%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th align="center">manifacturer</th>
<th align="center">n_of_snps</th>
<th align="center">name</th>
<th align="center">species</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">illumina</td>
<td align="center">54241</td>
<td align="center">IlluminaOvineSNP50</td>
<td align="center">Sheep</td>
</tr>
<tr class="even">
<td align="center">illumina</td>
<td align="center">605998</td>
<td align="center">IlluminaOvineHDSNP</td>
<td align="center">Sheep</td>
</tr>
<tr class="odd">
<td align="center">affymetrix</td>
<td align="center">56793</td>
<td align="center">AffymetrixAxiomOviCan</td>
<td align="center">Sheep</td>
</tr>
</tbody>
</table>
<p>Then collect all the SNPs for a certain chip by providing the SMARTER chip <em>name</em>. Please, consider that you will download more than 50K SNP for this chip and this will take a lot of time</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">variants</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_smarter_variants.html">get_smarter_variants</a></span><span class="op">(</span>
  species <span class="op">=</span> <span class="st">"Sheep"</span>,
  assembly <span class="op">=</span> <span class="st">"OAR3"</span>,
  query <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>
    chip_name <span class="op">=</span> <span class="st">"IlluminaOvineSNP50"</span>
  <span class="op">)</span>
<span class="op">)</span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Paolo Cozzi.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.3.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
